src/state.rs

```diff
@@ -109,6 +111,244 @@ pub struct ResolvedDns {
     dns_refresh_rate: std::time::Duration,
 }
 
+/// A SingleFlight can make sure that only the first request to really handle DNS resolving task
+/// for specified hostname and other requests in concurrency need to wait for the task completion.
+#[derive(serde::Serialize, Default, Debug)]
+struct SingleFlight {
+    // is_first_req_in can record requests number and it can be used to
+    // check the request number, such as the first request.
+    is_first_req_in: bool,
+    // wait_for_notification can help to implement the multiple waiters and a single notifier
+    // and it should be in Arc and protected by RWLock because of the concurrency envs.
+    #[serde(skip_serializing)]
+    wait_for_notification: Arc<Notify>,
+}
+
+impl Clone for SingleFlight {
+    fn clone(&self) -> Self {
+        SingleFlight {
+            is_first_req_in: self.is_first_req_in,
+            wait_for_notification: self.wait_for_notification.clone(),
+        }
+    }
+}
+
+impl SingleFlight {
+    fn new() -> Self {
+        SingleFlight {
+            // is_first_req_in is set to true because
+            // there always is the first request
+            is_first_req_in: true,
+            wait_for_notification: Arc::new(Notify::new()),
+        }
+    }
+
+    async fn wait_for_notifying(&self) {
+        self.wait_for_notification.notified().await;
+    }
+
+    fn notify_waiters(&self) {
+        self.wait_for_notification.notify_waiters();
+    }
+}
+
+impl DnsResolver {
+    pub fn new(dns_resolver_cfg: ResolverConfig, dns_resolver_opts: ResolverOpts) -> Self {
```

Comment:
remove `pub` on all methods? The struct is not public.

Reply:
okay, remove `pub`, done!